relay-server/src/metrics_extraction/metrics_summary.rs

```diff
@@ -0,0 +1,511 @@
+                };
+
+                let mut tags = BTreeMap::new();
+                if let Some(inner_tags) = summary.tags.value() {
+                    for (tag_key, tag_value) in inner_tags.iter() {
+                        let Some(tag_value) = tag_value.value() else {
+                            continue;
+                        };
+
+                        tags.insert(tag_key.clone(), tag_value.clone());
+                    }
+                }
+
+                let key = MetricsSummaryBucketKey {
+                    metric_name: MetricName::from(metric.clone()),
+                    tags,
+                };
+
+                let value = MetricsSummaryBucketValue::from_metric_summary(summary);
+
+                self.merge(key, value);
+            }
+        }
+    }
+
+    /// Merges a [`MetricsSummaryBucketKey`] and [`MetricsSummaryBucketValue`] in the spec.
+    fn merge(&mut self, key: MetricsSummaryBucketKey, value: MetricsSummaryBucketValue) {
+        match self.buckets.entry(key) {
+            Entry::Occupied(mut entry) => {
+                entry.get_mut().merge(value);
+            }
+            Entry::Vacant(entry) => {
+                entry.insert(value);
+            }
+        }
+    }
+
+    /// Applies the [`MetricsSummarySpec`] on a receiving [`Annotated<MetricsSummary>`].
+    ///
+    /// In case the [`MetricsSummarySpec`] is empty, no mutation will take place.
+    pub fn apply_on(mut self, receiver: &mut Annotated<MetricsSummary>) {
+        if self.buckets.is_empty() {
+            return;
+        }
+
+        if let Some(receiver) = receiver.value() {
+            self.merge_metrics_summary(receiver);
+        }
+
+        *receiver = self.into()
+    }
+}
+
+impl From<MetricsSummarySpec> for Annotated<MetricsSummary> {
+    /// Builds the [`MetricsSummary`] from the spec.
+    ///
+    /// Note that this method consumes the metrics_summary_spec itself, since the purpose of the metrics_summary_spec
+    /// is to be built and destroyed once the summary is ready to be computed.
+    fn from(spec: MetricsSummarySpec) -> Self {
+        let mut metrics_summary = BTreeMap::new();
+
+        for (key, value) in spec.buckets {
+            let tags = key
+                .tags
+                .into_iter()
+                .map(|(tag_key, tag_value)| (tag_key, Annotated::new(tag_value)))
+                .collect();
+
+            let metric_summary = MetricSummary {
+                min: value
+                    .min
+                    .map_or(Annotated::empty(), |m| Annotated::new(m.to_f64())),
+                max: value
+                    .max
+                    .map_or(Annotated::empty(), |m| Annotated::new(m.to_f64())),
+                sum: value
+                    .sum
+                    .map_or(Annotated::empty(), |m| Annotated::new(m.to_f64())),
+                count: Annotated::new(value.count),
+                tags: Annotated::new(tags),
+            };
```

Comment:
```suggestion
            let min = value.min.map_or(Annotated::empty(), |m| m.to_f64().into());
            let max = value.max.map_or(Annotated::empty(), |m| m.to_f64().into());
            let sum = value.sum.map_or(Annotated::empty(), |m| m.to_f64().into());
            let metric_summary = MetricSummary {
                min,
                max,
                sum,
                count: Annotated::new(value.count),
                tags: Annotated::new(tags),
            };
```

Not sure, feels like this is more readable?